lua/autorun/sh_cfc_spawnpoint.lua

```diff
@@ -0,0 +1,99 @@
+CreateConVar( "cfc_spawnpoints_interact_cooldown", 0.5, { FCVAR_ARCHIVE, FCVAR_REPLICATED }, "Per-player interaction cooldown for spawn points.", 0, 1000 )
+CreateConVar( "cfc_spawnpoints_health_max", 1500, { FCVAR_ARCHIVE, FCVAR_REPLICATED }, "Max health of spawnpoints. 0 to disable.", 0, 10000 )
+CreateConVar( "cfc_spawnpoints_health_regen", 200, { FCVAR_ARCHIVE, FCVAR_REPLICATED }, "Health regenerated per second by spawnpoints. 0 to disable.", 0, 10000 )
+CreateConVar( "cfc_spawnpoints_health_regen_cooldown", 10, { FCVAR_ARCHIVE, FCVAR_REPLICATED }, "If a spawnpoint takes damage, it must wait this long before it can start regenerating. 0 to disable.", 0, 10000 )
+
+local REMOVAL_WINDOW = CreateConVar( "cfc_spawnpoints_removal_window", 30, { FCVAR_ARCHIVE, FCVAR_REPLICATED }, "Player/point cooldowns only apply if a previous spawn point was removed in the past X seconds. 0 to not alter cooldowns.", 0, 1000 )
+
+
+----- GLOBAL FUNCTIONS -----
+
+--[[
+    - Determines whether or not a player is considered 'friendly' to a spawn point.
+        - i.e. they can link to it, if no cooldowns or other restrictions block them.
+    - Returns friendly, failReason
+    - You can override this function in InitPostEntity if you need a different 'friendliness' check.
+--]]
+function CFC_SpawnPoints.IsFriendly( spawnPoint, ply )
+    if not CPPI then
+        if spawnPoint:GetCreatingPlayer() == ply then return true end
+
+        return false, "You can only link to your own Spawn Points."
+    end
+
+    local owner = spawnPoint:CPPIGetOwner()
+    if ply == owner then return true end
+
+    local friends = owner.CPPIGetFriends and owner:CPPIGetFriends()
+
+    if not friends or friends == CPPI.CPPI_DEFER or friends == CPPI.CPPI_NOTIMPLEMENTED then
+        return false, "You can only link to your own Spawn Points."
+    end
+
+    if table.HasValue( friends, ply ) then return true end
+
+    return false, "You are not buddied with the Spawn Point's owner."
+end
+
+
+----- PRIVATE FUNCTIONS -----
+
+local function ignoreCooldownsDueToRemovalWindow( ply )
+    if REMOVAL_WINDOW:GetFloat() <= 0 then return false end
+
+    local lastRemoval = ply:GetNWFloat( "CFC_SpawnPoints_LastRemovedTime", 0 )
+    local hasBeenALongTime = CurTime() - lastRemoval > REMOVAL_WINDOW:GetFloat()
+
+    return hasBeenALongTime
+end
+
+
+----- SETUP -----
+
+hook.Add( "CFC_SpawnPoints_IgnorePlayerSpawnCooldown", "CFC_SpawnPoints_RemovalWindow", function( ply )
+    if ignoreCooldownsDueToRemovalWindow( ply ) then return true end
+end )
+
+hook.Add( "CFC_SpawnPoints_IgnorePointSpawnCooldown", "CFC_SpawnPoints_RemovalWindow", function( _spawnPoint, ply )
+    if ignoreCooldownsDueToRemovalWindow( ply ) then return true end
+end )
+
+hook.Add( "InitPostEntity", "CFC_SpawnPoints_BlockInvisibility", function()
+    local entityMeta = FindMetaTable( "Entity" )
+
+    local setColor = entityMeta.SetColor
+    function entityMeta:SetColor( color )
```

Comment:
this feels like a dirty + performance heavy way of doing this, i get that things like sf might be able to do it but i rather have sf being able to do it than preemptively detouring often ran functions